prov/sm2/src/sm2_coordination.h

```diff
@@ -48,7 +48,7 @@
 
 #include <rdma/providers/fi_prov.h>
 
-#define SM2_INJECT_SIZE	      4096
+#define SM2_INJECT_SIZE	      4048
```

Comment:
Can we calculate this size from SMR_ALIGN_BOUNDARY, like https://github.com/ofiwg/libfabric/blob/main/include/ofi_shm.h#L123

Reply:
### Option 1:
The `struct sm2_xfer_entry` is header + data.. this would require separating out the header which is a few hundered lines of code... which seems like overkill for a 1 line fix.  This is something that I want to do eventually... but I think it can wait.

### Option 2:
Change these to use SM2_INJECT_SIZE
```
└─╼ grep -rI "sizeof(struct sm2_xfer_entry)" prov/sm2
prov/sm2/src/sm2_util.c:	total_size += freestack_size(sizeof(struct sm2_xfer_entry),
prov/sm2/src/sm2_util.c:			   sizeof(struct sm2_xfer_entry));
```
and update:
```
struct sm2_xfer_entry {
	volatile long int next;
	uint64_t size;
	uint64_t cq_data;
	uint64_t tag;
	uint32_t op;
	uint32_t op_flags;
	uint32_t proto;
	sm2_gid_t sender_gid;
	uint8_t user_data[];
} __attribute__((aligned(16)));
```
### Option 3:
A good comment, and fix later, when we have a functional reason to change the header... I prefer option 3.

Reply:
I just tried to do option 1, and it doesn't work b/c we get padding:
```
#define SM2_INJECT_USER_DATA_LENGTH \
	(SM2_INJECT_SIZE - sizeof(struct sm2_inject_header))
...
struct sm2_inject_header {
	volatile long int next;
	uint64_t size;
	uint64_t cq_data;
	uint64_t tag;
	uint64_t context;
	uint32_t op;
	uint32_t op_flags;
	uint32_t proto;
	sm2_gid_t sender_gid;
} __attribute__((aligned(16)));

struct sm2_xfer_entry {
	struct sm2_inject_header hdr;
	uint8_t user_data[SM2_INJECT_USER_DATA_LENGTH];
} __attribute__((aligned(16)));
```

```
	printf("sm2_freestack_size: %d\n", SM2_INJECT_USER_DATA_LENGTH);
```

 sm2_freestack_size: 4032

Reply:
No but I don't think we need that b/c our memory performance is good, its the number of instructions on the critical path that is the problem.

Reply:
I'm confused. You're saying we're good with the memory utilization so we don't need it but you're also trying we need to align the freestack? Sounds like there are memory performance issues then?

Reply:
@a-szegel I'm not sure you want alignment.  I think you want to define that struct using `__attribute__((packed))` and then you can use the straight-forward computation of SM2_INJECT_SIZE as @shijin-aws suggests, right?

@a-szegel Can you expand on "memory performance is good"?  I think what you mean is that you are looking at two metrics: perhaps cache miss rate or page-faults, and instructions/cycle, and comparing SM2 to SHM as a baseline.  Is that correct?

Reply:
@lrbison 
1. Packing doesn't work as you expect, the inject size calculation is: SM2_INJECT_USER_DATA_LENGTH: 3992
2. From some benchmarking I did... it seemed like libfabric could gain a lot more performance by being algorithmically more efficient, rather than trying to be more memory efficient when compared to OMPI.

@aingerson 
1. This fix aims to improve page access via alignment.  You are suggesting that I improve cache hits via locality which is different than this fix... and I will look at in the future, but isn't in the same spirit as this fix.

Reply:
@a-szegel I was suggesting the improvement as both - the smr freestack code has all of the payload memory back to back chunks so that it could easily be aligned without the need for headers and such. This payload itself is still not aligned because of the header, and is going to span 2 pages because of the header. Separating it into inline + inject allows the inline protocol to only grab the single page and the inject protocol to grab 2 where one is entirely the payload

Reply:
On my other branch: https://github.com/ofiwg/libfabric/pull/8913, I lowered the inject size to 256 and 128 and didn't see an improvement in pt2pt latency... will do more tests with that.

Reply:
after doing more testing... it seems like the locality does provide a slight improvement on 96 rank alltoallw for small message sizes but doesn't impact latency... I can put out a separate fix